056-zk-kraft-migration.md

```diff
@@ -0,0 +1,451 @@
+The current reconciliation ends.
+
+**_Status changes_**
+
+If the migration finished, the FSM moves to the `DualWrite` state.
+If the migration is still ongoing, the FSM stays in the `Migration` state.
+
+#### Phase 3/4: Migrating brokers and controllers to full KRaft
+
+At the beginning of this phase, the user has the last opportunity to rollback the migration process.
+
+**_Preconditions_**
+
+The FSM is in the `DualWrite` state (as per `Kafka.status.migrationState` field).
+
+**_Trigger_**
+
+The user can rollback the process by applying the `strimzi.io/kraft: pre-migration` annotation on the `Kafka` custom resource.
+
+**_Actions_**
+
+In this case, the operator starts a reconciliation and, via the `KafkaReconciler` component, configures the brokers with no ZooKeeper migration flag, and no KRaft controllers quorum connection information.
+Finally, the operator rolls the brokers, via the `KafkaRoller`, and waits for them to be ready again.
+The current reconciliation ends.
+
+**_Status changes_**
+
+The FSM moves back to the `WaitBrokersInMigration`, so actually at the beginning of the Phase 2.
+
+---
+
+If the user doesn't rollback the migration, next step is to finalize the migration process. This step is irreversible.
+
+**_Preconditions_**
+
+The FSM is in the `DualWrite` state (as per `Kafka.status.migrationState` field) and the operator expects the user desire to finalize the migration, having a full KRaft based cluster and ZooKeeper out of the picture.
+
+**_Trigger_**
+
+The user applies the `strimzi.io/kraft: post-migration` annotation to the `Kafka` custom resource.
+
+**_Actions_**
+
+The operator starts a reconciliation and, via the `KafkaReconciler` component, updates the brokers' configuration by removing the ZooKeeper migration flag and connection to the ZooKeeper ensemble.
+It also has to replace the `broker.id` with `node.id`.
+More specifically, the brokers are automatically updated with the following configuration (the user cannot set them manually).
+
+```shell
+# replace broker.id with node.id
+node.id=0
+
+# Remove the migration enabled flag
+# zookeeper.metadata.migration.enable=true
+
+# Remove ZooKeeper client configuration
+# zookeeper.connect=<connection-to-zookeeper>
+
+# Don't set the IBP, KRaft uses "metadata.version" feature flag
+# inter.broker.protocol.version=3.5
+```
+
+The operator rolls the brokers, via the `KafkaRoller`, and waits for them to be ready again.
+They are now connected to the KRaft controllers quorum only and not to the ZooKeeper ensemble anymore.
+
+When brokers are rolled and ready again, the operator updates the controllers' configuration by removing the ZooKeeper migration flag and connection to the ZooKeeper ensemble.
+More specifically, the controllers are automatically updated with the following configuration (the user cannot set them manually).
```

Comment:
Does it have to follow this order? That makes it dependent on KafkaRoller I guess. Should there be two internal states instead? First one to remove Zoo from brokers and second to remve Zoo from controllers?

Reply:
Yes this is the order as per official Kafka procedure.
At the beginning the FSM had a state in the middle as you are suggesting but I thought that it was not necessary.
Isn't the KafkaRoller able to roll brokers first and controllers then all in one reconcile?

Reply:
Well, I guess what the reconciler will do depends on the other proposal. This is why I raised it here because you are relying on something that is also a work in progress.

Also, to be clear -> I do not care if the state machine has 1 or 20 states. The way you move between the states - requiring the user to approve - is what worries me. So the thought here in this comment was to first roll the brokers, move the state and in next recon roll the controllers. But without any user interaction.

Reply:
> Well, I guess what the reconciler will do depends on the other proposal. This is why I raised it here because you are relying on something that is also a work in progress.

Just to be sure are you referring to the "KRaft Support in Kafka Roller" proposal?

Reply:
Yes.